SU2_CFD/src/solvers/CSolver.cpp

```diff
@@ -2065,79 +2066,81 @@ void CSolver::AdaptCFLNumber(CGeometry **geometry,
     /* Check whether we achieved the requested reduction in the linear
      solver residual within the specified number of linear iterations. */
 
-    bool reduceCFL = false;
-    su2double linResFlow = solverFlow->GetResLinSolver();
-    su2double linResTurb = -1.0;
-    if ((iMesh == MESH_0) && (config->GetKind_Turb_Model() != NONE)) {
-      linResTurb = solverTurb->GetResLinSolver();
-    }
+    su2double linResTurb = 0.0;
+    if ((iMesh == MESH_0) && solverTurb) linResTurb = solverTurb->GetResLinSolver();
 
-    su2double maxLinResid = max(linResFlow, linResTurb);
-    if (maxLinResid > 0.5) {
-      reduceCFL = true;
-    }
+    /* Max residual and iterations between flow and turbulence. */
+    const su2double linRes = max(solverFlow->GetResLinSolver(), linResTurb);
+
+    /* Tolerance limited to a "reasonable" value. */
+    const su2double linTol = max(0.001, config->GetLinear_Solver_Error());
 
     /* Check that we are meeting our nonlinear residual reduction target
-     over time so that we do not get stuck in limit cycles. */
+     over time so that we do not get stuck in limit cycles, this is done
+     on the fine grid and applied to all others. */
 
     SU2_OMP_MASTER
     { /* Only the master thread updates the shared variables. */
 
-    Old_Func = New_Func;
-    unsigned short Res_Count = 100;
-    if (NonLinRes_Series.size() == 0) NonLinRes_Series.resize(Res_Count,0.0);
+    /* Check if we should decrease or if we can increase, the 20% is to avoid flip-flopping. */
+    resetCFL = linRes > 1.0;
+    reduceCFL = linRes > 1.2*linTol;
+    canIncrease = linRes < linTol;
 
-    /* Sum the RMS residuals for all equations. */
+    if ((iMesh == MESH_0) && (Res_Count > 0)) {
+      Old_Func = New_Func;
+      if (NonLinRes_Series.empty()) NonLinRes_Series.resize(Res_Count,0.0);
 
-    New_Func = 0.0;
-    for (unsigned short iVar = 0; iVar < solverFlow->GetnVar(); iVar++) {
-      New_Func += solverFlow->GetRes_RMS(iVar);
-    }
-    if ((iMesh == MESH_0) && (config->GetKind_Turb_Model() != NONE)) {
-      for (unsigned short iVar = 0; iVar < solverTurb->GetnVar(); iVar++) {
-        New_Func += solverTurb->GetRes_RMS(iVar);
-      }
-    }
+      /* Sum the RMS residuals for all equations. */
 
-    /* Compute the difference in the nonlinear residuals between the
-     current and previous iterations. */
+      New_Func = 0.0;
+      for (unsigned short iVar = 0; iVar < solverFlow->GetnVar(); iVar++) {
+        New_Func += log10(solverFlow->GetRes_RMS(iVar));
```

Comment:
I think it makes sense to use the log of the residual as the orders of magnitude can differ significantly.